specs/error-handling-guideline.md

```diff
@@ -0,0 +1,159 @@
+- Recommended solution is required and should follow the general guiding principles described above.
+
+### Examples
+
+#### Sign an artifact without artifact reference or signing key
+
+Current behavior and output:
+
+```
+$ notation sign --signature-format cose
+Error: missing reference
+```
+
+Suggested error message:
+
+```
+$ notation sign --signature-format cose
+Error: missing artifact reference required for signing. 
+Please specify an artifact reference. Run "notation sign --help" for more options and examples.
+```
+
+#### Sign an artifact with an non-existing signing key in a key vault
+
+Current behavior and output:
+
+```
+$ notation sign localhost:5000/test-repo:v1  --signature-format cose --plugin wabbitnetworks-kv --id https://feynman-kv.vault.wabbit.net/keys/feynmankv-networks-io/6670ffa5cb694c49b1e0a6bb6bdefaaa
+Warning: Always sign the artifact using digest(@sha256:...) rather than a tag(:v1) because tags are mutable and a tag reference can point to a different artifact than the one signed.
+Error: describe-key command failed: ERROR: A certificate with (name/id) feynmankv-networks-io/versions/6670ffa5cb694c49b1e0a6bb6bdefaaa was not found in this key vault. If you recently deleted this certificate you may be able to recover it using the correct recovery command. For help resolving this issue, please see https://go.wabbit.net/fwlink/?linkid=2125182
+Status: 404 (Not Found)
+ErrorCode: CertificateNotFound
+
+Content:
+{"error":{"code":"CertificateNotFound","message":"A certificate with (name/id) feynmankv-networks-io/versions/6670ffa5cb694c49b1e0a6bb6bdefaaa was not found in this key vault. If you recently deleted this certificate you may be able to recover it using the correct recovery command. For help resolving this issue, please see https://go.wabbit.net/fwlink/?linkid=2125182"}}
+
+Headers:
+Cache-Control: no-cache
+Pragma: no-cache
+x-ms-keyvault-region: eastus
+x-ms-client-request-id: a2923244-ed47-461b-9dc1-d0b9f4202788
+x-ms-request-id: 96103d99-c372-449f-adba-8d24b7d5da7e
+x-ms-keyvault-service-version: 1.9.1116.1
+x-ms-keyvault-network-info: conn_type=Ipv4;addr=20.65.162.175;act_addr_fam=InterNetwork;
+X-Content-Type-Options: REDACTED
+Strict-Transport-Security: REDACTED
+Date: Wed, 13 Dec 2023 07:27:33 GMT
+Content-Length: 376
+Content-Type: application/json; charset=utf-8
+Expires: -1
+```
+
+Suggested error message:
+
+```
+$ notation sign localhost:5000/test-repo:v1  --signature-format cose --plugin wabbitnetworks-kv --id https://feynman-kv.vault.wabbit.net/keys/feynmankv-networks-io/6670ffa5cb694c49b1e0a6bb6bdefaaa
+Warning: Always sign the artifact using digest(@sha256:...) rather than a tag(:v1) because tags are mutable and a tag reference can point to a different artifact than the one signed.
+ERROR: A certificate with (name/id) feynmankv-networks-io/versions/6670ffa5cb694c49b1e0a6bb6bdefaaa was not found in this key vault. 
```

Comment:
In this example, users used the command correctly, but the key did not exist in the key vault. The error is from server, so maybe we should make it clear error is from server side, not from client side, and the suggested error message also missed the response code from server.

Reply:
I mark this as resolved due to https://github.com/notaryproject/notation/pull/834#issuecomment-1882649172